static/js/components/ProfileSelect.tsx

```diff
@@ -0,0 +1,168 @@
+}
+
+const ProfileSelect: FC<Props> = ({ notify, selected, setSelected }) => {
+  const toggleProfile = (item: LxdProfile) => {
+    const newSelection = selected.includes(item.name)
+      ? selected.filter((i) => i !== item.name)
+      : [...selected, item.name];
+    setSelected(newSelection);
+  };
+
+  const moveUp = (item: LxdProfile) => {
+    const index = selected.indexOf(item.name);
+    if (index === 0) {
+      return;
+    }
+    const newSelection = [...selected];
+    newSelection.splice(index, 1);
+    newSelection.splice(index - 1, 0, item.name);
+    setSelected(newSelection);
+  };
+
+  const moveDown = (item: LxdProfile) => {
+    const index = selected.indexOf(item.name);
+    if (index === selected.length) {
+      return;
+    }
+    const newSelection = [...selected];
+    newSelection.splice(index, 1);
+    newSelection.splice(index + 1, 0, item.name);
+    setSelected(newSelection);
+  };
+
+  const { data: projectProfiles = [], error } = useQuery({
+    queryKey: [queryKeys.profiles],
+    queryFn: fetchProfiles,
+  });
+
+  if (error) {
+    notify.failure("Could not load profiles.", error);
+  }
+
+  const unselected = projectProfiles.filter(
+    (profile) => !selected.includes(profile.name)
+  );
+
+  return (
+    <>
+      <Label className="u-sv-1">Apply profiles</Label>
+      <div className="u-text--muted u-sv1">
+        Add profiles and create a hierarchy
+      </div>
+      <Row className="p-profile-select">
+        <Col size={6} className="p-profile-select-box">
+          <i className="p-profile-select-title">Available profiles</i>
+          <hr />
+          <div className="p-profile-select-list">
+            <List
+              items={unselected.map((profile) => (
+                <Fragment key={profile.name}>
+                  <span
+                    className="p-profile-select-name u-truncate"
+                    title={profile.name}
+                  >
+                    {profile.name}
+                  </span>
+                  <div className="u-float-right">
+                    <Button
+                      dense
+                      hasIcon
+                      small
+                      type="button"
+                      onClick={() => toggleProfile(profile)}
+                    >
+                      Apply
+                    </Button>
+                  </div>
+                </Fragment>
+              ))}
+            />
+          </div>
+        </Col>
+        <Col size={6} className="p-profile-select-box">
+          <i className="p-profile-select-title">Applied profiles</i>
+          <hr />
+          <div className="p-profile-select-list">
+            <List
+              items={selected.map((name, i) => {
+                const profile = projectProfiles.find(
+                  (profile) => profile.name === name
+                );
+                if (!profile) {
+                  return null;
+                }
```

Comment:
Having `selected` as a string array is causing many unnecessary checks and type inconsistency between the `selected` and the `unselected` array - wouldn't it be better to have both as `LxdProfile` arrays instead? And then you map the array in the parent (`InstanceFormGuided`).

Reply:
I had the full profiles in the state before, but the problem is the initial state. It should be only the profile with name "default", and when I use strings I can just hardcode the name as init state. If it would be the object, I'd need to fetch it in the parent and select it.

Reply:
Good point - hadn't noticed that. But is it safe to assume that a profile called `default` always exists? Because currently what you are doing is basically adding it blindly to the `selected` list with no checks about its existence. So, what I propose to do instead, is to move the assignment of the `default` profile to the `ProfileSelect` component, where you would need to fetch all the profiles anyway, and there, after checking that a profile called `default` exists, you add it to the `selected` list by default.

Reply:
Yeah it exists always and is always called default. There is no way to ask the default profile for a project. Because it is alwas a profile called default. You can't remove or rename it or set a new default profile.

Reply:
See the discussion at https://chat.canonical.com/canonical/pl/kkyq6d1zq3nadbae6nzgpiz1yr

Reply:
Then it's something that needs to be addressed, because currently you could delete the default profile. I just submitted https://github.com/canonical/lxd-ui/pull/117 for that.

Back on topic, I would propose one of the following:
- Moving the default assignment logic to the child (`ProfileSelect`)
- Keep the logic in the parent but create a mock `LxdProfile` object with only the name set to `default`

And then have `selected` as a `LxdProfile` array.

Reply:
Thanks for the #117 PR!

I think keeping the selected profiles as a string array with the names is reducing complexity.

Reply:
Ok, LGTM then!